ffi/orcjit.cpp

```diff
@@ -0,0 +1,337 @@
+#include "core.h"
+#include "llvm-c/LLJIT.h"
+#include "llvm-c/Orc.h"
+#include <sstream>
+
+#include "llvm/AsmParser/Parser.h"
+#include "llvm/ExecutionEngine/Orc/Core.h"
+#include "llvm/ExecutionEngine/Orc/ExecutionUtils.h"
+#include "llvm/ExecutionEngine/Orc/LLJIT.h"
+#include "llvm/ExecutionEngine/Orc/RTDyldObjectLinkingLayer.h"
+#include "llvm/ExecutionEngine/SectionMemoryManager.h"
+#include "llvm/IRReader/IRReader.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include "llvm/Support/SourceMgr.h"
+using namespace llvm;
+using namespace llvm::orc;
+
+inline LLJIT *unwrap(LLVMOrcLLJITRef P) { return reinterpret_cast<LLJIT *>(P); }
+
+inline TargetMachine *unwrap(LLVMTargetMachineRef TM) {
+    return reinterpret_cast<TargetMachine *>(TM);
+}
+
+inline LLVMOrcJITTargetMachineBuilderRef wrap(JITTargetMachineBuilder *JTMB) {
+    return reinterpret_cast<LLVMOrcJITTargetMachineBuilderRef>(JTMB);
+}
+
+class JITDylibTracker {
+  public:
+    std::shared_ptr<LLJIT> lljit;
+    IntrusiveRefCntPtr<llvm::orc::ResourceTracker> tracker;
+    JITDylib &dylib;
+    JITDylibTracker(std::shared_ptr<LLJIT> &lljit_, JITDylib &dylib_,
+                    IntrusiveRefCntPtr<llvm::orc::ResourceTracker> &&tracker_)
+        : lljit(lljit_), dylib(dylib_), tracker(tracker_) {}
+};
+
+typedef struct {
+    uint8_t element_kind;
+    char *value;
+    size_t value_len;
+} LinkElement;
+typedef struct {
+    char *name;
+    uint64_t address;
+} SymbolAddress;
+extern "C" {
+
+API_EXPORT(std::shared_ptr<LLJIT> *)
+LLVMPY_CreateLLJITCompiler(LLVMTargetMachineRef tm, const char **OutError) {
+    LLJITBuilder builder;
+
+    if (tm) {
+        // The following is based on
+        // LLVMOrcJITTargetMachineBuilderCreateFromTargetMachine. However,
+        // we can't use that directly because it destroys the target
+        // machine, but we need to keep it alive because it is referenced by
+        // / shared with other objects on the Python side.
+        auto *template_tm = unwrap(tm);
+
+        builder.setJITTargetMachineBuilder(
+            JITTargetMachineBuilder(template_tm->getTargetTriple())
+                .setCPU(template_tm->getTargetCPU().str())
+                .setRelocationModel(template_tm->getRelocationModel())
+                .setCodeModel(template_tm->getCodeModel())
+                .setCodeGenOptLevel(template_tm->getOptLevel())
+                .setFeatures(template_tm->getTargetFeatureString())
+                .setOptions(template_tm->Options));
+    }
+    builder.setObjectLinkingLayerCreator(
+        [&](llvm::orc::ExecutionSession &session, const llvm::Triple &triple)
+            -> std::unique_ptr<llvm::orc::ObjectLayer> {
+            auto linkingLayer =
+                std::make_unique<llvm::orc::RTDyldObjectLinkingLayer>(
```

Comment:
This is not using JITLink, which we will want an option to use.

See https://github.com/JuliaLang/julia/pull/43664/files#diff-b858d0b6a127703fdaa4091f393fff5033565d010102e00b4426363ee7dfa91dR15-R17

Reply:
JITLink makes enabling debugging very difficult. I haven't been able to get it to work on Linux. Julia's solution to the debugging is to do it manually, and they don't support MacOS. It seems like this gets easier in future LLVM version as there's just a Boolean flag to enable debugging.